pkg/aws/ec2/ec2.go

```diff
@@ -473,38 +491,59 @@ func (e *EC2) CreateNewInstanceSG(sgName *string, sgID *string, vpcID *string) (
 	return oInstanceSG.GroupId, nil
 }
 
-// GetVPCID retrieves the VPC that the subnets passed are contained in.
+// GetVPCID returns the VPC of the instance the controller is currently running on.
+// This is achieved by getting the identity document of the EC2 instance and using
+// the DescribeInstances call to determine its VPC ID.
 func (e *EC2) GetVPCID(subnets []*string) (*string, error) {
 	var vpc *string
 
-	if len(subnets) == 0 {
-		return nil, fmt.Errorf("Empty subnet list provided to getVPCID")
-	}
-
-	key := fmt.Sprintf("%s-vpc", *subnets[0])
+	// If previously looked up (and not expired) the VpcId will be stored in the cache under the
+	// key 'vpc'.
+	key := "vpc"
 	item := e.cache.Get(key)
 
-	if item == nil {
-		subnetInfo, err := e.DescribeSubnets(&ec2.DescribeSubnetsInput{
-			SubnetIds: subnets,
-		})
-		if err != nil {
-			return nil, err
-		}
+	// cache hit: return (pointer of) VpcId value
+	if item != nil {
+		vpc = item.Value().(*string)
+		albprom.AWSCache.With(prometheus.Labels{"cache": "vpc", "action": "hit"}).Add(float64(1))
+		return vpc, nil
+	}
 
-		if len(subnetInfo.Subnets) == 0 {
-			return nil, fmt.Errorf("DescribeSubnets returned no subnets")
-		}
+	// cache miss: begin lookup of VpcId based on current EC2 instance
+	// retrieve identity of current running instance
+	identityDoc, err := EC2Metadatasvc.GetInstanceIdentityDocument()
+	if err != nil {
+		return nil, err
+	}
 
-		vpc = subnetInfo.Subnets[0].VpcId
-		e.cache.Set(key, vpc, time.Minute*60)
+	// capture instance ID for lookup in DescribeInstances
+	// don't bother caching this value as it should never be re-retrieved unless
+	// the cache for the VpcId (looked up below) expires.
+	descInstancesInput := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(identityDoc.InstanceID)},
+	}
 
-		albprom.AWSCache.With(prometheus.Labels{"cache": "vpc", "action": "miss"}).Add(float64(1))
-	} else {
-		vpc = item.Value().(*string)
-		albprom.AWSCache.With(prometheus.Labels{"cache": "vpc", "action": "hit"}).Add(float64(1))
+	// capture description of this instance for later capture of VpcId
+	descInstancesOutput, err := e.DescribeInstances(descInstancesInput)
+	if err != nil {
+		return nil, err
+	}
+
+	// Before attempting to return VpcId of instance, ensure at least 1 reservation and instance
+	// (in that reservation) was found.
+	if len(descInstancesOutput.Reservations) < 1 {
+		return nil, fmt.Errorf("When looking up VPC ID could not identify instance. Found %d reservations"+
+			" in AWS call. Should have found atleast 1.", len(descInstancesOutput.Reservations))
+	}
+	if len(descInstancesOutput.Reservations[0].Instances) < 1 {
+		return nil, fmt.Errorf("When looking up VPC ID could not identify instance. Found %d instances"+
+			" in AWS call. Should have found atleast 1.", len(descInstancesOutput.Reservations))
 	}
 
+	vpc = descInstancesOutput.Reservations[0].Instances[0].VpcId
```

Comment:
@joshrosso isn't good check for nil or blank?

Otherwise the logic looks good.

Reply:
@jescarri I believe i'm already doing this above. 

https://github.com/coreos/alb-ingress-controller/pull/297/files/653c431f3cdb5a316dd69a8ed249695898e4d0ad#diff-15b7ef6a224ea1db965f77e6d392cfadR532

^ can you verify this is an adequate nil check you're thinking of?

Reply:
yep, that one looks for nil or len == 0 on the output, however  no testing of ` descInstancesOutput.Reservations[0].Instances[0].VpcId` != nil or != "", just an extra safeguard, in the event aws returning "" for vpcID.